### ISSUE TITLE
preserve environment variables using dockerTools 

### ISSUE THREAD

**Issue by srghma** [NixOS/nixpkgs#37172]
[Question in stackoverflow](https://stackoverflow.com/questions/49321680/dockertools-buildimage-removes-environment-variables-of-fromimage-image-how-to)

I expect [this variables][1] to be preserved when I use `dockerTools.buildImage`

    $ docker run -it lnl7/nix:latest env
    PATH=/root/.nix-profile/bin:/run/current-system/sw/bin
    HOSTNAME=52be36e10f96
    TERM=xterm
    MANPATH=/root/.nix-profile/share/man:/run/current-system/sw/share/man
    NIX_PATH=nixpkgs=/nix/store/d7r0qlms9w7k0m5kmvinp52abxbncafv-nixpkgs-unstable-2018-01-13
    GIT_SSL_CAINFO=/nix/store/dblcx21khkizv3h9n0haqbkmzig3n3wd-nss-cacert-3.33/etc/ssl/certs/ca-bundle.crt
    SSL_CERT_FILE=/nix/store/dblcx21khkizv3h9n0haqbkmzig3n3wd-nss-cacert-3.33/etc/ssl/certs/ca-bundle.crt
    HOME=/root
    

But using this

    $ cat runner/nixops_runner.nix
    { pkgs ? import <nixpkgs> {} }:
    
    with pkgs;
    
    let
      workdir = "/app";
    
      sshConfig = writeText "ssh_config" ''
        Host *
        StrictHostKeyChecking=no
        UserKnownHostsFile /dev/null
        IdentityFile /ssh/id_rsa
      '';
    
      nixFromDockerHub = dockerTools.pullImage {
        imageName = "lnl7/nix";
        imageTag = "latest";
    
        # this hash will need change if the tag is updated at docker hub
        sha256 = "031xba6yi72v87v560wqbcsi80hqbwx498qz5b5jxigj36yb3k36";
      };
    in
    
    dockerTools.buildImage {
      name = "nixops_runner";
    
      fromImage = nixFromDockerHub;
    }
    

I get this
    
    $ docker load --input $(nix-build runner/nixops_runner.nix --no-out-link) && docker run -it nixops_runner env
    Loaded image: nixops_runner:latest
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
    HOSTNAME=9d673e788746
    TERM=xterm
    HOME=/root

How to preserve environment variables or image?

  [1]: https://github.com/LnL7/nix-docker/blob/93d877076e8ca67273a77f0bec69b931811649c8/default.nix#L65

**Comment by coretemp** [comment]
@BjornMelgaard Is there documentation which says that it should do that? If not, then this is just a feature request/question.

**Comment by BjornMelgaard** [comment]
@coretemp then yes. I just expect it to work as Dockerfile.

**Comment by coretemp** [comment]
@BjornMelgaard Not sure what you mean. Do you mean "Yes, this is a feature request and there is no basis in the documentation under the nixos.org domain suggesting that this should work."?

**Comment by BjornMelgaard** [comment]
Yes, this is a feature request and there is no basis in the documentation under the nixos.org domain suggesting that this should work )

**Comment by BjornMelgaard** [comment]
this is the very important feature for me. I want to refactor my configurations to use nix instead of dockerfiles (because nix is more user-friendly) and I can't image how to I will do this if `pullImage` is not preserving environment variables

**Comment by BjornMelgaard** [comment]
@coretemp what would you suggest to make this happen?

**Comment by thefloweringash** [comment]
This is now fixed on master (#87253).

<details>
  <summary>Verification</summary>

```nix
{ pkgs ? import <nixpkgs> {} }:

with pkgs;

let
  nixFromDockerHub = dockerTools.pullImage {
    imageName = "lnl7/nix";
    imageDigest = "sha256:4d588295b59eda18215b11dad9e03a08773b7452ebca4c5d4627c8f9eea1b2d4";
    sha256 = "1ca3sbazy7r3abjacr3wjfwpa6bgbh0pdva0apjwcnxf4jmyag7d";
  };
in

dockerTools.buildImage {
  name = "nixops_runner";

  fromImage = nixFromDockerHub;
}
```

```
$ skopeo inspect docker-archive:$(nix-build nixops_runner.nix) | jq -r '.Env[]'
PATH=/root/.nix-profile/bin:/run/current-system/sw/bin
MANPATH=/root/.nix-profile/share/man:/run/current-system/sw/share/man
NIX_PAGER=cat
NIX_PATH=nixpkgs=/nix/store/z124xpfhlsshxs775jrc8nd83chqkbzy-nixpkgs-20.09pre228453.dcb64ea42e6
NIX_SSL_CERT_FILE=/nix/store/dsgqkz2rmcy8as7kb25i8gwni0kkif4z-nss-cacert-3.52/etc/ssl/certs/ca-bundle.crt
```

```
$ skopeo inspect docker://lnl7/nix | jq -r '.Env[]'
PATH=/root/.nix-profile/bin:/run/current-system/sw/bin
MANPATH=/root/.nix-profile/share/man:/run/current-system/sw/share/man
NIX_PAGER=cat
NIX_PATH=nixpkgs=/nix/store/z124xpfhlsshxs775jrc8nd83chqkbzy-nixpkgs-20.09pre228453.dcb64ea42e6
NIX_SSL_CERT_FILE=/nix/store/dsgqkz2rmcy8as7kb25i8gwni0kkif4z-nss-cacert-3.52/etc/ssl/certs/ca-bundle.crt
```

</details>
